ui/src/logs/utils/matchSections.ts

```diff
@@ -0,0 +1,52 @@
+import {MatchType, Filter, MatchSection} from 'src/types/logs'
+import uuid from 'uuid'
+
+export const getMatchSections = (
+  filters: Filter[],
+  text: string
+): MatchSection[] => {
+  if (filters.length === 0) {
+    return [createSection(MatchType.NONE, text)]
+  }
+
+  try {
+    const pattern = filtersToPattern(filters)
+    return sectionOnPattern(pattern, text)
+  } catch (e) {
+    console.error('Syntax Error: bad search filter expression')
+
+    return [createSection(MatchType.NONE, text)]
+  }
+}
+
+const sectionOnPattern = (pattern: RegExp, text) => {
+  let sections = []
+  let remaining = text
+
+  for (
+    let match = remaining.match(pattern);
+    match !== null;
+    match = remaining.match(pattern)
+  ) {
+    remaining = match[match.length - 1]
+    sections = [
+      ...sections,
```

Comment:
I would prefer to see pushing onto the existing array rather than creating a whole new array here simply because this is going to cause a lot of churn in GC for no visible immutability win.